cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizer.java

```diff
@@ -0,0 +1,236 @@
+import io.cdap.cdap.proto.id.ArtifactId;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpRequestConfig;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.apache.twill.filesystem.LocalLocationFactory;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.validation.constraints.NotNull;
+
+public class ArtifactLocalizer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ArtifactLocalizer.class);
+
+  // TODO: Move this directory name into a cConf property
+  private static final String PD_DIR = "data/";
+  private final RemoteClient remoteClient;
+  private final RetryStrategy retryStrategy;
+  private final LocationFactory locationFactory;
+
+  @Inject
+  public ArtifactLocalizer(DiscoveryServiceClient discoveryServiceClient, LocationFactory locationFactory) {
+    this.remoteClient = new RemoteClient(discoveryServiceClient, Constants.Service.APP_FABRIC_HTTP,
+                                         HttpRequestConfig.DEFAULT,
+                                         Constants.Gateway.INTERNAL_API_VERSION_3);
+    this.locationFactory = locationFactory;
+    this.retryStrategy = RetryStrategies.exponentialDelay(100, 1000, TimeUnit.MILLISECONDS);
+  }
+
+  @VisibleForTesting
+  public ArtifactLocalizer(DiscoveryServiceClient discoveryServiceClient, File basePath) {
+    this.remoteClient = new RemoteClient(discoveryServiceClient, Constants.Service.APP_FABRIC_HTTP,
+                                         HttpRequestConfig.DEFAULT,
+                                         Constants.Gateway.INTERNAL_API_VERSION_3);
+    locationFactory = new LocalLocationFactory(basePath);
+    this.retryStrategy = RetryStrategies.exponentialDelay(100, 1000, TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Gets the location on the local filesystem for the given artifact. This method handles fetching the artifact as well
+   * as caching it.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch
+   * @return The Local Location for this artifact
+   * @throws ArtifactNotFoundException if the given artifact does not exist
+   * @throws IOException if there was an exception while fetching or caching the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public Location getArtifact(ArtifactId artifactId) throws Exception {
+    LOG.debug("Fetching artifact info for {}", artifactId);
+    Location localLocation = getArtifactDirLocation(artifactId);
```

Comment:
I believe this class is always using local disk for caching fetched artifact, right? Why bother to use Location instead of using `File` or `Path` directly?

Reply:
The classes that rely on this method (ex. `InMemoryConfigurator`, `RemotePluginFinder` etc.) use Location as the interface. I could update this to return File or Path but it would need to be converted to location anyway further up in the call stack.

Reply:
Spoke with Terence offline and agreed that we should refactor everything to use `Path` or `File` and only convert it when we need to